# Software for high-quality proteome selection using OMArk

Genome alignments and comparisons are commonly used in current genomic studies. They rely on the increasing quantity of available genomic sequences and annotations available on multiple databases. The amount of heterogeneous data forces researchers to carefully select the data they use to find a balance between data quality and taxonomic coverage of the chosen genomes.

To help researchers choose among all the available data, we developed software that selects species of interest by combining their diversity and genome quality. Using the pre-computed output of OMArk and the NCBI ultrametric tree, our algorithm evaluates the overall quality of the genomes, compares the quality of the source databases, and proposes a list of selected species according to the user's configuration.

Using selected Vertebrate species, we illustrate the functionality of the developed software and its outcomes. We highlighted the importance of the choice of the source database and proposed a new tool to select genomic data for species.
